xdsl/ir.py

```diff
@@ -1000,25 +1126,45 @@ def insert_op(self,
         Insert one or multiple operations at a given index in the block.
         The operations should not be attached to another block.
         """
-        if index < 0 or index > len(self.ops):
+
+        if index < 0 or index > self.len_ops():
             raise ValueError(
                 f"Can't insert operation in index {index} in a block with "
-                f"{len(self.ops)} operations.")
+                f"{self.len_ops()} operations.")
         if not isinstance(ops, list):
             ops = [ops]
-        if name:
-            for curr_op in ops:
-                for res in curr_op.results:
-                    res.name = name
-        for op in ops:
-            self._attach_op(op)
-        self.ops = self.ops[:index] + ops + self.ops[index:]
+
+        first_op = self.first_op
+
+        if first_op is None:
+            if index:
+                raise ValueError(
+                    f"Can't insert operation in index {index} in a block with "
+                    f"no operations.")
+
+            self.add_ops(ops)
+            return
+
+        i = 0
```

Comment:
I feel this part is missing some comments.
Like, what is `i`?

Reply:
Updated with new logic and more comments